### PR TITLE
Remove unused forward declarations

### PR DESCRIPTION
**Pull request by drwells**
This set of commits improves the handling of forward declarations versus headers. Unfortunately it conflicts with #2331 (both remove some extra headers): I don't mind merging that one first and then fixing this.

I propose that we add something like
```
There are several places in deal.II where we use forward declarations in header
files. The reason for this is that we can, hopefully, improve compilation speeds
by not using headers when we just need to mark a certain type as an argument to
a function. The convention used in deal.II is that, if all we need is a type
name, then the type may be forward declared in a header; if a function (or
member function) can return a value then the declaration of that value's type
should be available by including that header. For example, dofs/dof_handler.h
includes dofs/dof_accessor.h so that one can run

dof_handler.begin_active()->is_active()

without explicitly including the header for the type returned by begin_active().
```
to the style guide. Thoughts?

### CODE REVIEW

File: include/deal.II/base/function.templates.h [hunk 1]
```diff
@@ -26,11 +26,6 @@
 
 DEAL_II_NAMESPACE_OPEN
 
-
-template <int dim, typename Number>
-const unsigned int Function<dim, Number>::dimension;
```
Comment (bangerth) [inline include/deal.II/base/function.templates.h]: Oh, this is not a declaration but a definition. I don't think you want to remove it.

File: include/deal.II/dofs/dof_handler.h [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include <deal.II/base/iterator_range.h>
 #include <deal.II/base/std_cxx11/shared_ptr.h>
 #include <deal.II/dofs/block_info.h>
+#include <deal.II/dofs/dof_accessor.h>
```
Comment (bangerth) [inline include/deal.II/dofs/dof_handler.h]: Do you really need this in the header file? Do we have inline functions where we actually dereference iterators in the header file?

**Comment by bangerth** [comment]
Let's go with your patch first -- I guess I can always see what breaks when I try to rebase mine.

Can you think about my comments and see what needs to be done? I'm ok with the addition to the style guide, though don't really think it is necessary.

**Comment by bangerth** [comment]
You do have a thing for "boring and repetitive work" (your words, IIRC). It is very much appreciated -- this is the stuff that nobody would do if you didn't do it!

File: include/deal.II/base/function.templates.h (hunk 1 above)
Comment (drwells) [inline include/deal.II/base/function.templates.h]: Isn't the definition/definition of this already in `function.h`? That is why I removed this.

File: include/deal.II/dofs/dof_handler.h (hunk 2 above)
Comment (drwells) [inline include/deal.II/dofs/dof_handler.h]: This one was pretty hard to figure out. I think the answer is 'yes' (and this is what inspired the proposed blurb on style) but the template errors I got without this were unusually hard to understand.

**Comment by drwells** [comment]
Thank you for the kind words (I have a strange sense of fun). I should probably sleep on this before addressing more comments.

File: include/deal.II/base/function.templates.h (hunk 1 above)
Comment (bangerth) [inline include/deal.II/base/function.templates.h]: The *declaration* is there. Also the *initialization*. But this is the place where you *define* it, i.e., where the compiler allocates space for the variable.

File: include/deal.II/dofs/dof_handler.h (hunk 2 above)
Comment (bangerth) [inline include/deal.II/dofs/dof_handler.h]: OK. I guess most people will use both DoFHandler and the iterators, so there's likely little harm.

File: include/deal.II/base/function.templates.h (hunk 1 above)
Comment (drwells) [inline include/deal.II/base/function.templates.h]: You are correct: we cannot use it as an lvalue unless we do this. Today I learned something new about this corner of C++ :)

I suppose I am used to (being new to C++) declaring these things as `constexpr` and ignoring such details.

For posterity: it is always good to refer to the master:

http://www.stroustrup.com/bs_faq2.html#in-class

**Comment by drwells** [comment]
I addressed the review comments (and added the blurb in the style guide).

**Comment by tamiko** [comment]
:-)

**Pull request merged**